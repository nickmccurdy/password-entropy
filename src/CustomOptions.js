import Options from './Options'
import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

export default class CustomOptions extends PureComponent {
  static propTypes = {
    onChange: PropTypes.func.isRequired
  }

  state = {
    wordlistLength: 7776
  }

  handleChange = event => this.setState({ wordlistLength: parseInt(event.target.value, 10) })
  possiblePasswords = () => this.state.wordlistLength

  render () {
    return (
      <Options possiblePasswords={this.possiblePasswords} {...this.props}>
        <h3>Diceware</h3>
        <label>
          <h4>Wordlist Length</h4>
          <input value={this.state.wordlistLength} onChange={this.handleChange} type="number" min="1" required/>
        </label>
      </Options>
    )
  }
}
