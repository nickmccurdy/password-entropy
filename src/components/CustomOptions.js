import handleChange from '../handleChange'
import Options from './Options'
import React, { Component } from 'react'

export default class CustomOptions extends Component {
  static propTypes = Options.sharedPropTypes
  static shortName = 'Custom'

  state = {
    possibleItems: 1
  }

  handleChange = handleChange.bind(this)
  possiblePasswords = () => this.state.possibleItems

  render () {
    return (
      <Options possiblePasswords={this.possiblePasswords} {...this.props}>
        <h3>Custom</h3>
        <label>
          <h4>Possible Items</h4>
          <input name="possibleItems" value={this.state.possibleItems} onChange={this.handleChange} type="number" min="1" required/>
        </label>
      </Options>
    )
  }
}
