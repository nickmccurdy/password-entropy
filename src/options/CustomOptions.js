import { setCustom } from '../actions'
import FormGroup from '../ui/FormGroup'
import PropTypes from 'prop-types'
import { path } from 'ramda'
import React, { PureComponent } from 'react'
import { FormControl } from 'react-bootstrap'
import { connect } from 'react-redux'

export class CustomOptions extends PureComponent {
  static propTypes = {
    possibleItems: PropTypes.number.isRequired,
    setCustom: PropTypes.func.isRequired
  }

  render () {
    return (
      <div>
        <h3>Custom</h3>

        <FormGroup id="possibleItems" label="Possible Items" icon="question-circle">
          <FormControl value={this.props.possibleItems} onChange={this.props.setCustom} type="number" min="0" required/>
        </FormGroup>
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  possibleItems: path(['options', 'custom'], state)
})

export default connect(mapStateToProps, { setCustom })(CustomOptions)
