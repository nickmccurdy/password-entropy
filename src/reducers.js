import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
