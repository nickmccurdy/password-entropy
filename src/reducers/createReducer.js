import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
