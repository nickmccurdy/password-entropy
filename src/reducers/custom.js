import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
