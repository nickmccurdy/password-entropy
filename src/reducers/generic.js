import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
