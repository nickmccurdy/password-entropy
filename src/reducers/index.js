import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
