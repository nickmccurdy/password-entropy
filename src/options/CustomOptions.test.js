import { shallow } from 'enzyme'
import React from 'react'
import { createStore } from 'redux'

import { CustomOptions, mapStateToProps } from './CustomOptions'
import reducers from '../reducers'

test('CustomOptions', () => {
  const setCustom = jest.fn()
  const wrapper = shallow(<CustomOptions possibleItems={1} setCustom={setCustom} />)
  wrapper.find('FormControl').simulate('change')

  expect(wrapper).toMatchSnapshot()
  expect(setCustom).toHaveBeenCalled()
})

test('mapStateToProps', () => {
  const state = createStore(reducers).getState()
  expect(mapStateToProps(state)).toEqual({ possibleItems: 0 })
})
