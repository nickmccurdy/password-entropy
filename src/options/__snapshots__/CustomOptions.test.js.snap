// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomOptions 1`] = `
<FormGroup
  icon="question-circle"
  id="custom"
  label="Possible Items"
>
  <FormControl
    bsClass="form-control"
    componentClass="input"
    min="0"
    onChange={[Function]}
    required={true}
    type="number"
    value={1}
  />
</FormGroup>
`;
